Executing DetachedCriteria with subcriteria multiple times fails

### Background

For a filtered view with pagination we use a DetachedCriteria which specifies all data to be returned. We then run `criteria.count()` and `criteria.list(params)` to receive a page. However this fails, if the criteria contains a detached criteria as subcriteria.

### Steps to Reproduce

Have a look at the attached project ([gorm-bug-report](https://github.com/timaebi/gorm-bug-report)) which describes the error. There is one integration test, which currently fails but shouldn't in our opinion. The project setup was done in a straight forward manner based on grails 3.3.11:

1. `grails create-app gorm-bug-report`
2. create domain classes `User` and `Organisation` with a many-to-one relationship
3. create a test
4. run `./gradlew integrationTest` and see the test failing

### Expected Behaviour

Executing the same DetachedCriteria multiple times should return the same results and not throw an exception.

### Actual Behaviour

The first execution of the query works well, however a second call throws the following exception:

```
Expected no exception of type 'org.springframework.core.convert.ConverterNotFoundException' to be thrown, but got it nevertheless
	at spock.lang.Specification.notThrown(Specification.java:107)
	at AssociationSubCriteriaSpec.$tt__$spock_feature_0_0(AssociationSubCriteriaSpec.groovy:49)
	at AssociationSubCriteriaSpec.test run detached criteria with association criteria multiple times_closure2(AssociationSubCriteriaSpec.groovy)
	at groovy.lang.Closure.call(Closure.java:420)
	at groovy.lang.Closure.call(Closure.java:436)
	at grails.gorm.transactions.GrailsTransactionTemplate$1.doInTransaction(GrailsTransactionTemplate.groovy:68)
	at org.springframework.transaction.support.TransactionTemplate.execute(TransactionTemplate.java:133)
	at grails.gorm.transactions.GrailsTransactionTemplate.executeAndRollback(GrailsTransactionTemplate.groovy:65)
	at AssociationSubCriteriaSpec.test run detached criteria with association criteria multiple times(AssociationSubCriteriaSpec.groovy)
Caused by: org.springframework.core.convert.ConverterNotFoundException: No converter found capable of converting from type [org.hibernate.criterion.DetachedCriteria] to type [bug.report.Organisation]
	at org.springframework.core.convert.support.GenericConversionService.handleConverterNotFound(GenericConversionService.java:324)
	at org.springframework.core.convert.support.GenericConversionService.convert(GenericConversionService.java:206)
	at org.springframework.core.convert.support.GenericConversionService.convert(GenericConversionService.java:187)
	at org.grails.orm.hibernate.query.AbstractHibernateQuery.doTypeConversionIfNeccessary(AbstractHibernateQuery.java:277)
	at org.grails.orm.hibernate.query.AbstractHibernateQuery.add(AbstractHibernateQuery.java:170)
	at org.grails.datastore.gorm.finders.DynamicFinder.applyDetachedCriteria(DynamicFinder.java:601)
	at grails.gorm.DetachedCriteria.withPopulatedQuery_closure8(DetachedCriteria.groovy:756)
	at groovy.lang.Closure.call(Closure.java:420)
	at groovy.lang.Closure.call(Closure.java:436)
	at org.grails.datastore.gorm.GormStaticApi.withDatastoreSession_closure24(GormStaticApi.groovy:862)
	at groovy.lang.Closure.call(Closure.java:420)
	at org.grails.datastore.mapping.core.DatastoreUtils.execute(DatastoreUtils.java:319)
	at org.grails.datastore.gorm.AbstractDatastoreApi.execute(AbstractDatastoreApi.groovy:40)
	at org.grails.datastore.gorm.GormStaticApi.withDatastoreSession(GormStaticApi.groovy:861)
	at grails.gorm.DetachedCriteria.withPopulatedQuery(DetachedCriteria.groovy:740)
	at grails.gorm.DetachedCriteria.list(DetachedCriteria.groovy:135)
	at grails.gorm.DetachedCriteria.list(DetachedCriteria.groovy:134)
	at AssociationSubCriteriaSpec.$tt__$spock_feature_0_0(AssociationSubCriteriaSpec.groovy:46)
	... 7 more
```

### Environment Information

- **Operating System**: macOS 10.15.3
- **GORM Version:** 6.1.12.RELEASE
- **Grails Version (if using Grails):** 3.3.11
- **JDK Version:** Java 8

### Example Application

The [gorm-bug-report](https://github.com/timaebi/gorm-bug-report) repository contains an integration test, which fails because of the bug.

### Related Issues

This issue might be related to grails/grails-data-mapping#59 .

--- a/grails-datastore-gorm-hibernate5/src/main/groovy/org/grails/orm/hibernate/query/AbstractHibernateQuery.java
+++ b/grails-datastore-gorm-hibernate5/src/main/groovy/org/grails/orm/hibernate/query/AbstractHibernateQuery.java
@@ -164,10 +164,10 @@ public abstract class AbstractHibernateQuery extends Query {
             Object value = pc.getValue();
             if (value instanceof QueryableCriteria) {
                 setDetachedCriteriaValue((QueryableCriteria) value, pc);
-            }
-            // ignore Size related constraints
-            else {
-                doTypeConversionIfNeccessary(getEntity(), pc);
+            } else {
+                if (!(value instanceof DetachedCriteria)) {
+                    doTypeConversionIfNeccessary(getEntity(), pc);
+                }
             }
         }
         if (criterion instanceof DetachedAssociationCriteria) {
@@ -266,6 +266,7 @@ public abstract class AbstractHibernateQuery extends Query {
 
     @SuppressWarnings("unchecked")
     static void doTypeConversionIfNeccessary(PersistentEntity entity, PropertyCriterion pc) {
+        // ignore Size related constraints
         if (pc.getClass().getSimpleName().startsWith(SIZE_CONSTRAINT_PREFIX)) {
             return;
         }

--- a/grails-datastore-gorm-hibernate5/src/test/groovy/grails/gorm/tests/DetachCriteriaSubquerySpec.groovy
+++ b/grails-datastore-gorm-hibernate5/src/test/groovy/grails/gorm/tests/DetachCriteriaSubquerySpec.groovy
@@ -9,7 +9,7 @@ class DetachCriteriaSubquerySpec extends GormDatastoreSpec {
 
     @Override
     List getDomainClasses() {
-        return [User, Group, GroupAssignment]
+        return [User, Group, GroupAssignment, Organisation]
     }
 
     void "test detached associated criteria in subquery"() {
@@ -43,6 +43,30 @@ class DetachCriteriaSubquerySpec extends GormDatastoreSpec {
         result.size() == 1
     }
 
+    void "test executing detached criteria in sub-query multiple times"() {
+
+        setup:
+        Organisation orgA = new Organisation(name: "A")
+        orgA.addToUsers(email: 'user1@a')
+        orgA.addToUsers(email: 'user2@a')
+        orgA.addToUsers(email: 'user3@a')
+        orgA.save(flush: true)
+        Organisation orgB = new Organisation(name: "B")
+        orgB.addToUsers(email: 'user1@b')
+        orgB.addToUsers(email: 'user2@b')
+        orgB.save(flush: true)
+
+        when:
+        DetachedCriteria<User> criteria = User.where {
+            inList('organisation', Organisation.where { name == 'A' || name == 'B' }.id())
+        }
+        List<User> result = criteria.list()
+        result = criteria.list()
+
+        then:
+        result.size() == 5
+    }
+
     void "test that detached criteria subquery should create implicit alias instead of using this_"() {
 
         setup:
@@ -74,9 +98,11 @@ class DetachCriteriaSubquerySpec extends GormDatastoreSpec {
     }
 
     private User createUser(String email) {
-        User user = new User()
-        user.email = email
-        user.save(flush: true)
+        User user = new User(email: email)
+        Organisation defaultOrg = Organisation.findOrCreateByName("default")
+        defaultOrg.addToUsers(user)
+        defaultOrg.save(flush: true)
+        user
     }
 
     private Group createGroup(String name, User supervisor) {
@@ -99,6 +125,7 @@ class DetachCriteriaSubquerySpec extends GormDatastoreSpec {
 @Entity
 class User implements HibernateEntity<User> {
     String email
+    static belongsTo = [organisation: Organisation]
     static mapping = {
         table 'T_USER'
     }
@@ -120,4 +147,10 @@ class GroupAssignment implements HibernateEntity<GroupAssignment> {
     static mapping = {
         table 'T_GROUP_ASSIGNMENT'
     }
+}
+
+@Entity
+class Organisation implements HibernateEntity<Organisation> {
+    String name
+    static hasMany = [users: User]
 }

